006-tls-encrypt-the-kafka-connect-rest-api.md

```diff
@@ -0,0 +1,98 @@
+# TLS encrypting the Kafka Connect REST API
+
+This proposal reduces access to the Kafka Connect REST API by enabling TLS encryption on the Kafka Connect REST listener when the connectors running on the Kafka Connect cluster are managed by the `KafkaConnector` or `KafkaMirrroMaker2` operators.
+
+## Current situation
+
+Currently, instances of Kafka Connect that are deployed by the Strimzi operators are configured with the default REST API endpoint settings.
+This means that the Kafka Connect REST API endpoint uses HTTP on port 8083 and that the Strimzi `KafkaConnect`, `KafkaConnector` and `KafkaMirrorMaker2` operators make unsecured REST client calls based on this default configuration.
+
+The default network policies created by the Strimzi operators restrict incoming REST API calls to only allow access from the operator pod.
+Users can define further network policies can be provided to override the default policy and allow wider access.
+
+## Motivation
+
+There is currently no way to TLS encrypt the Kafka Connect REST API, leaving the Connect REST listener as perhaps the last key endpoint that cannot currently be restricted using Strimzi.
+
+## Proposal
+
+This proposal changes the behaviour of the `KafkaMirrorMaker2` operator and the `KafkaConnect` operator.
+In the `KafkaConnect` operator case, the behaviour is only changed when the `strimzi.io/use-connector-resources: "true"` annotation is applied in the `KafkaConnect` CR (meaning that all connectors running on the Connect cluster are managed by the `KafkaConnector` operator).
+
+The default behaviour of the `KafkaMirrorMaker2` operator and the `KafkaConnect` operator will be to deploy Kafka Connect clusters with a single TLS encrypted REST API listener.
+A self-signed TLS certificate and key required to enable TLS encryption on the REST API listener are generated by the operator, stored in secrets for use by the operator clients and the Kafka Connect pods.
```

Comment:
Why is the TLS certificate self-signed? Can't we use the cluster CA (which is self-signed cluster root certificate) to sign the TLS certificate for each Kafka Connect worker node as we do for each Kafka broker?
In this way, the client (so the operator calling the REST API) che verify the TLS certificate using the cluster CA certificate.

Reply:
The instance of Kafka Connect that the operator is calling may not be connected to a Strimzi Kafka - it can be any Kafka instance, so there might not be a cluster CA available to use. 

We may be able to use the cluster CA if it is available (in the same namespace, etc) or we could generate a Connect-specific CA for this proposal - do you think that generating signed TLS certificates from a CA cert would be better than generating self-signed certificates directly?

Reply:
Tbh I see a self-signed certificate less secure. Self-signed certs are used for well-recognized CAs which then signs other certificates (or it could be a trusted internal CA in your organization). Are we sure that we want to support an external Kafka Connect not deployed by Strimzi? I thought that this PR was about securing the Kafka Connect instance deployed by Strimzi.

Reply:
@ppatierno - Sorry I wasn't clear - I meant an external (non-Strimzi) Kafka cluster that a `KafkaConnect` CR is configured to use. This proposal is about the REST API communication between the operator and a Strimzi-managed Kafka Connect. 

As the cluster CA is only created when you are using a Kafka cluster created via a `Kafka` CR, the cluster CA may not be available for use by the Strimzi-managed Kafka Connect instance. 

We could generate a connect-cluster-CA for sole use by the Kafka Connect instance, and generate certificates using that - would that be your preference here?

Reply:
I agree that the Kafka Cluster CA cannot be used as a basis here because Connect is not bound to any particular Kafka CR.

I can see the concerns about the self-signed term. But the Kafka Cluster CA is also self-signed really. So I'm not sure I can see it make any difference here.

Reply:
Ok but should at this point allow users to specify their own Kafka Connect CA as we do for the Kafka Cluster CA?

Reply:
If we don't use a CA we cannot let users provide custom CA - that would be only server certificate I guess. Me personally, I'm fine to wait with this if there will be any interest. The Connect API is different from the Kafka listeners so the demand might be different as well.

Reply:
I was misleading here. I thought the idea was to generate a self-signed Kafka Connect CA used for signing the worker nodes certificates but actually it's just a self-signed certificate (not a CA).